lapack/gonum/dpbtrf.go

```diff
@@ -0,0 +1,188 @@
+// Copyright ©2019 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gonum
+
+import (
+	"gonum.org/v1/gonum/blas"
+	"gonum.org/v1/gonum/blas/blas64"
+)
+
+func (impl Implementation) Dpbtrf(uplo blas.Uplo, n, kd int, ab []float64, ldab int) (ok bool) {
+	const nbmax = 32
+
+	switch {
+	case uplo != blas.Upper && uplo != blas.Lower:
+		panic(badUplo)
+	case n < 0:
+		panic(nLT0)
+	case kd < 0:
+		panic(kdLT0)
+	case ldab < kd+1:
+		panic(badLdA)
+	}
+
+	// Quick return if possible.
+	if n == 0 {
+		return true
+	}
+
+	if len(ab) < (n-1)*ldab+kd {
+		panic(shortAB)
+	}
+
+	opts := string(blas.Upper)
+	if uplo == blas.Lower {
+		opts = string(blas.Lower)
+	}
+	nb := impl.Ilaenv(1, "DPBTRF", opts, n, kd, -1, -1)
+	// The block size must not exceed the semi-bandwidth kd, and must not
+	// exceed the limit set by the size of the local array work.
+	nb = min(nb, nbmax)
+
+	if nb <= 1 || kd < nb {
+		// Use unblocked code.
+		return impl.Dpbtf2(uplo, n, kd, ab, ldab)
+	}
+
+	// Use blocked code.
+	ldwork := nb
+	work := make([]float64, nb*ldwork)
+	bi := blas64.Implementation()
+	if uplo == blas.Upper {
+		// Compute the Cholesky factorization of a symmetric band
+		// matrix, given the upper triangle of the matrix in band
+		// storage.
+
+		// Process the band matrix one diagonal block at a time.
+		for i := 0; i < n; i += nb {
+			ib := min(nb, n-i)
+			// Factorize the diagonal block.
+			ok := impl.Dpotf2(uplo, ib, ab[i*ldab:], ldab-1)
+			if !ok {
+				return false
+
+			}
+			if i+ib >= n {
+				continue
+			}
+			// Update the relevant part of the trailing submatrix.
+			// If A11 denotes the diagonal block which has just been
+			// factorized, then we need to update the remaining
+			// blocks in the diagram:
+			//
+			//  A11   A12   A13
+			//        A22   A23
+			//              A33
+			//
+			// The numbers of rows and columns in the partitioning
+			// are ib, i2, i3 respectively. The blocks A12, A22 and
+			// A23 are empty if ib = kd. The upper triangle of A13
+			// lies outside the band.
+			i2 := min(kd-ib, n-i-ib)
+			if i2 > 0 {
+				// Update A12.
+				bi.Dtrsm(blas.Left, blas.Upper, blas.Trans, blas.NonUnit, ib, i2,
+					1, ab[i*ldab:], ldab-1, ab[i*ldab+ib:], ldab-1)
```

Comment:
The absence of `kd` is row versus column major?

Reply:
Yes. There are two ways to translate Fortran banded to Go banded code:
1. The manual one: In Fortran, in case of Upper, the first index select the diagonal. `kd+1` is the main diagonal, `kd+1-1` is the first super-diagonal, etc. In general, `kd+1-d` is the d-th super-diagonal. So if we rewrite the row index in this form, we get our column index. For example, `1+IB = KD+1+IB-KD = KD+1-(KD-IB)`, so our column index would be `d = kd-ib`. To get the row index, we must subtract `d` from the Fortran column index. For example, `I+KD-d = I+KD-(KD-IB) = I+IB`. So the final index is `(i+ib)*ldab+kd-ib`.
2. The semi-manual one: In case of Upper, take Fortran code for Lower, swap row/column indices and convert to 0-based.

2 is of course less work but you have take care of sizes, sidedness, and transpose, which must match the actual uplo, not the uplo from which you are taking the indices. Overall, it's quite unpleasant and annoying.